src/core/thread.d

```diff
@@ -631,7 +633,8 @@ class Thread
         }
 
         // Start thread as non-suspendable, undone in thread_entryPoint
-        atomicStore(m_isInCriticalRegion, true);
+        synchronized (Thread.criticalRegionLock)
+            m_isInCriticalRegion = true;
```

Comment:
I'm not sure setting isInCriticalRegion here is a good idea to begin with. As mentioned in the comment above, thread creation during DLL startup can easily create dead locks, e.g. a new thread might be blocked by the OS lock on DllMain.